src/mip/coefficient_strengthening.jl

```diff
@@ -0,0 +1,178 @@
+
+In particular, we consider the i-th constraints of the form
+aᵢ x ⩽ bᵢ and where xⱼ is a integer/binary variable.
+If aᵢⱼ ⩾ d = bᵢ - Mᵢⱼ - aᵢⱼ (uⱼ - 1) > 0
+where Mᵢⱼ is maximal activity of the i-th row without considering xⱼ
+and uⱼ is upper bound of xⱼ, then we transform
+aᵢⱼ <- aᵢⱼ - d
+bᵢ <- bᵢ - duⱼ .
+
+A similar update rule is applied for the constraints of the form cᵢ ⩽ aᵢ x.
+
+In case the constraint is ranged, i.e, cᵢ ⩽ aᵢ x ⩽ bᵢ, no coefficient strengthening is performed.
+
+The new coefficients are stored by updating ps.pb0 .
+If some coefficient is reduced to 0, it is kept as 0 in both ps.pb0.arows and ps.pb0.acols,
+and the number of nonzeros in each row/column is updated.
+
+If a row/column is reduced to 0 after this step, it will be set to be inactive.
+"""
+
+function maximal_activity(ps::PresolveData{T}, i::Int)::T where {T}
+    sup = zero(T)
+
+    row = ps.pb0.arows[i]
+    for (j, aij) in zip(row.nzind, row.nzval)
+        ps.colflag[j] || continue
+        if aij > zero(T)
+            sup += aij * ps.ucol[j]
+        elseif aij < zero(T)
+            sup += aij * ps.lcol[j]
+        end
+        isfinite(sup) || return sup
+    end
+    return sup
+end
+
+function minimal_activity(ps::PresolveData{T}, i::Int)::T where {T}
+    inf = zero(T)
+
+    row = ps.pb0.arows[i]
+    for (j, aij) in zip(row.nzind, row.nzval)
+        ps.colflag[j] || continue
+        if aij > zero(T)
+            inf += aij * ps.lcol[j]
+        elseif aij < zero(T)
+            inf += aij * ps.ucol[j]
+        end
+        isfinite(inf) || return inf
+    end
+    return inf
+end
+
+function upperbound_strengthening(ps::PresolveData{T}, i::Int, j::Int, coef, max_act::T) where {T}
+    # perform coef strengthening for one constraint of the form a'x <= u
+    new_bound = ps.urow[i]
+    new_coef = coef
+    if coef > 0
+        d = new_bound - max_act - coef * (ps.ucol[j]-1)
+        if coef >= d > 0
+            new_coef = new_coef - d
+            new_bound -= d*ps.ucol[j]
+        end
+    elseif coef < 0
+        d = new_bound - max_act - coef * (ps.lcol[j]+1)
+        if -coef >= d > 0
+            new_coef = new_coef + d
+            new_bound += d*ps.lcol[j]
+        end
+    end
+    return new_coef, new_bound
+end
+
+function lowerbound_strengthening(ps::PresolveData{T}, i::Int, j::Int, coef, min_act::T) where {T}
+    # perform coef strengthening for one constraint of the form l < = a'x
+    new_bound = ps.lrow[i]
+    new_coef = coef
+    if coef > 0
+        d = -new_bound + min_act + coef*(ps.lcol[j]+1)
+        if coef >= d > 0
+            new_coef = coef - d
+            new_bound -= d*ps.lcol[j]
+        end
+    elseif coef < 0
+        d = -new_bound + min_act + coef*(ps.ucol[j]-1)
+        if -coef >= d > 0
+            new_coef = coef + d
+            new_bound += d*ps.ucol[j]
+        end
+    end
+    return new_coef, new_bound
+end
+
+function zero_coefficient_strengthening!(ps::PresolveData{T}) where {T}
```

Comment:
I'm not sure about this naming. It sounds like you're only applying coefficient strengthening on zero coefficients (maybe?). I would just call this `_coefficient_strengthenening!`.